src/transformStateWithClones.js

```diff
@@ -6,8 +6,38 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const resultStates = [];
+
+  let stateCopy = Object.assign({}, state);
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties': {
+        Object.assign(stateCopy, action.extraData);
+        break;
+      }
+
+      case 'removeProperties': {
+        for (const key of action.keysToRemove) {
+          delete stateCopy[key];
+        }
+        break;
+      }
+
+      case 'clear': {
+        stateCopy = {};
+        break;
+      }
+
+      default:
+        continue;
+    }
```

Comment:
The use of a `switch` statement here is appropriate as it handles a limited set of conditions. However, the `default` case should be used for error handling, not for continuing the loop. If an unknown action type is encountered, it would be better to throw an error or handle this case explicitly.